lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -1617,6 +1618,183 @@ FullApplySite SILCombiner::rewriteApplyCallee(FullApplySite apply,
+      if (!info)
+        continue;
+      stringLiterals.push_back(info.getValue().inst);
+      stringCompares.push_back(apply);
+      subject = apply->getArgument(1);
+    }
+  }
+
+  if (stringLiterals.empty())
+    return false;
+
+  SmallVector<ValueDecl *, 1> results;
+  // We'll let ObjectOutliner::replaceFindStringCall do the heavy lifting.
+  Builder.getASTContext().lookupInSwiftModule("_findStringSwitchCase", results);
+  if (results.size() != 1)
+    return false;
+
+  auto *funcDecl = dyn_cast<FuncDecl>(results.front());
+  if (!funcDecl)
+    return false;
+
+  std::string mangled = SILDeclRef(funcDecl, SILDeclRef::Kind::Func).mangle();
+  SILFunction *replacementFunc =
+      module.findFunction(mangled, SILLinkage::PublicExternal);
+  if (!replacementFunc)
+    return false;
+
+  results.pop_back();
+  Builder.getASTContext().lookupInSwiftModule("_allocateUninitializedArray",
+                                              results);
+  if (results.size() != 1)
+    return false;
+
+  funcDecl = dyn_cast<FuncDecl>(results.front());
+  if (!funcDecl)
+    return false;
+
+  mangled = SILDeclRef(funcDecl, SILDeclRef::Kind::Func).mangle();
+  // TODO: this only exists when there is an enum in the source file.
+  SILFunction *makeArray =
+      module.findFunction(mangled, SILLinkage::PublicExternal);
+  if (!makeArray)
+    return false;
+
+  results.pop_back();
+  // Look up "assert" to get the StaticString type.
+  Builder.getASTContext().lookupInSwiftModule("assert", results);
+  if (results.size() != 1)
+    return false;
+
+  funcDecl = dyn_cast<FuncDecl>(results.front());
+  if (!funcDecl)
+    return false;
+
+  mangled = SILDeclRef(funcDecl, SILDeclRef::Kind::Func).mangle();
+  SILFunction *assertFn =
+      module.findFunction(mangled, SILLinkage::PublicExternal);
+  if (!assertFn)
+    return false;
+
+  // General variables we'll need later.
+  auto dummyLoc = subject.getLoc();
+  auto wordTy = SILType::getBuiltinWordType(Builder.getASTContext());
+  auto byteTy = SILType::getBuiltinIntegerType(8, Builder.getASTContext());
+  auto boolTy = SILType::getBuiltinIntegerType(1, Builder.getASTContext());
+  auto i64Ty = SILType::getBuiltinIntegerType(64, Builder.getASTContext());
+  auto assertFnFuncTy = assertFn->getLoweredFunctionType();
+  auto staticStringBaseTy =
+      assertFnFuncTy->getParameters()[2].getSILStorageType(module,
+                                                           assertFnFuncTy);
+  auto staticStringTy = staticStringBaseTy.getObjectType();
+  auto staticStringAddrTy = staticStringBaseTy.getAddressType();
+  auto replacementFuncTy = replacementFunc->getLoweredFunctionType();
+  auto staticStringArrayTy = replacementFuncTy->getParameters()[0]
+                                 .getSILStorageType(module, replacementFuncTy)
+                                 .getAddressType();
+
+  // Move all the strings to the top of the function.
+  Builder.setInsertionPoint(stringLiterals.front());
```

Comment:
The basic problem here is that you are scanning the function in arbitrary order (there is no guarantee of any basic block ordering). The first found string literal does not necessarily dominate all other string literals, e.g.
```
let subject = some_string
if (some_other_condition) {
  if (subject == "abc") { ... } // does not dominate "def"
} else {
  if (subject == "def") { ... }  // does not dominate "abc"
}
```

You need to take the control flow / dominance into account.

Reply:
Hmm, you're right. What do you think the best way to resolve this would be? I'm worried about making this function too slow. I think I'm going to run one pass to check if the function has an optimizable block in it. If it does, I'll just keep running the current code until the array isn't empty, otherwise, I'll return `false`.

Reply:
Actually, this won't happen. [The loop](https://github.com/apple/swift/blob/7bf0dd68d50b6f6859ae5bdf407c38867c603aae/lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp#L1629-L1652) looks for apply instructions and will only collect the string literal if it's one of the arguments. 

This optimization would optimize `A` and not `B` but, it would also get `B` in the next pass. 
```swift
func (x: String, y: String) -> Int {
// A
  if x == "a" { return 1 } 
// B
  switch y {
    case "a": return 1
    case "b": return 2
    // ...
  }
}
```
The example you gave _will_ get optimized.

Reply:
It's not about if it gets optimized, it's about correctness.
Maybe I misunderstand something here, but in my example, if you set the Builder insertion point to the first comparison and you generate the call at this place, it will not be executed if some_other_condition is false. But in the else-branch you will use the result of the call (actually the Verifier will already complain at compile time).

Reply:
To fix this, you would want to make sure that one comparison dominates all other comparisons (using the DominanceAnalysis).
You could also calculate the common dominator of all comparisons or use the definition point of 'subject'. But this is problematic because it could call _findStringSwitchCase even if no comparison is executed in the original code, which might degrade performance. For example:
```
let subject = some_string
// Inserting _findStringSwitchCase() here is not good, because we don't know if any of condition1 or condition2 is true. So we might end up calling _findStringSwitchCase whereas the original code does basically nothing.
if (condition1) {
  if (subject == "abc") { ... } // does not dominate "def"
}
if (condition2) {
  if (subject == "def") { ... }  // does not dominate "abc"
}
```

Reply:
I see what you're saying. My misunderstanding. I'll look into fixing this.

Reply:
@eeckstein Fixed based on your suggestion (using `DominanceAnalysis`). Thanks for the examples and for explaining.